friktion/friktion_swap_client/swap.py

```diff
@@ -0,0 +1,521 @@
+from .friktion_anchor.instructions import create, exec, exec_msg
+from solana.rpc.async_api import AsyncClient
+from solana.system_program import SYS_PROGRAM_ID
+from .swap_order_template import SwapOrderTemplate
+from .bid_details import BidDetails
+from spl.token.constants import ASSOCIATED_TOKEN_PROGRAM_ID, TOKEN_PROGRAM_ID
+from solana.transaction import Transaction
+from enum import Enum
+import asyncio
+from spl.token.async_client import AsyncToken
+
+
+def get_token_account(token_account_pk: PublicKey):
+    http_client = Client(commitment=Processed)
+    resp = http_client.get_account_info(token_account_pk)
+    account_data = layouts.ACCOUNT_LAYOUT.parse(decode_byte_string(resp["result"]["value"]["data"][0]))
+    return account_data
+
+class Network(Enum):
+    DEVNET = 1
+    TESTNET = 2
+    MAINNET = 3
+
+def get_url_for_network(network: Network) -> str:
+    if network == Network.DEVNET:
+        return "https://api.devnet.solana.com"
+    elif network == Network.TESTNET:
+        return "https://api.testnet.solana.com"
+    else:
+        # mainnet
+        return "https://solana-api.projectserum.com"
+
+MIN_REQUIRED_ALLOWANCE = 100
+
+# ---------------------------------------------------------------------------
+# Swap Program
+# ---------------------------------------------------------------------------
+class SwapContract():
+
+    network: Network
+    url: str
+    
+    def __init__(self, network: Network):
+        self.network = network
+        self.url = get_url_for_network(network)
+
+
+    # delegate should be the PDA of the swap order
+    def give_allowance(self, wallet: Wallet, token_acct_to_delegate: PublicKey, mint: PublicKey, allowance: int) -> bool:
+
+        client = Client(self.url)
+        token = Token(
+            client,
+            mint,
+            TOKEN_PROGRAM_ID,
+            wallet.payer
+        )
+
+        acct_info = token.get_account_info(token_acct_to_delegate)
+        # use a global authority for all swaps
+        # assert acct_info.owner == swap_order.counterparty
+        
+        token.approve(
+            token_acct_to_delegate,
+            DELEGATE_AUTHORITY_ADDRESS,
+            wallet.public_key,
+            allowance
+        )
+
+        return True
+    
+    # delegate should be the PDA of the swap order
+    def verify_allowance(self, wallet: Wallet, mint: PublicKey, token_account: PublicKey, with_assertions: bool) -> Tuple[int, int]:
+        client = Client(self.url)
+        token = Token(
+            client,
+            mint,
+            TOKEN_PROGRAM_ID,
+            wallet.payer
+        )
+        
+        acct_info = token.get_account_info(token_account)
+        print(acct_info.delegated_amount)
+
+        if with_assertions:
+            assert acct_info.delegate == DELEGATE_AUTHORITY_ADDRESS
+            assert acct_info.delegated_amount >= MIN_REQUIRED_ALLOWANCE
+
+        return (acct_info.delegated_amount, acct_info.amount)
```

Comment:
Noticed it does not need the `wallet.payer` to get the account info, so we may be able to simplify this method interface:
```suggestion
    # We may be able to simplify verify_allowance
    def verify_allowance(self, mint: PublicKey, token_account: PublicKey) -> bool:
        client = Client(self.url)
        payer = None # not needed
        token = Token(client, mint, TOKEN_PROGRAM_ID, payer)
        acct_info = token.get_account_info(token_account)
        return acct_info.delegate == DELEGATE_AUTHORITY_ADDRESS and acct_info.delegated_amount >= MIN_REQUIRED_ALLOWANCE
```

Reply:
this looks great! changing now